tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -172,11 +171,14 @@ step :: (Carrier sig m, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m Bool
 step = ask >>= liftIO . ts_tree_cursor_goto_next_sibling
 
 -- | Run an action over the children of the current node.
-push :: (Carrier sig m, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a -> m a
+push :: (Carrier sig m, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a -> m (Maybe a)
 push m = do
-  void $ ask >>= liftIO . ts_tree_cursor_goto_first_child
-  a <- m
-  a <$ (ask >>= liftIO . ts_tree_cursor_goto_parent)
+  hasChildren <- ask >>= liftIO . ts_tree_cursor_goto_first_child
```

Comment:
`ts_tree_cursor_goto_first_child` is documented to return `false` when the node has no children, rather than e.g. advancing the cursor to a position with no current node. Thus, we must only run the action when it returns `true`, which in turn means we need to return the action’s result in `Maybe`.